File: modules/foundation/src/main/java/com/projecthub/base/task/domain/validation/TaskValidator.java
package com.projecthub.base.task.domain.validation;

import com.projecthub.base.shared.exception.ValidationException;
import com.projecthub.base.task.domain.entity.Task;
import com.projecthub.base.task.domain.value.TaskValue;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.Set;

@Slf4j
@Component
public class TaskValidator {

    public void validateTasks(final Set<Task> tasks) {
        TaskValidator.log.debug("Validating {} tasks", tasks.size());
        tasks.forEach(this::validateTask);
    }

    private void validateTask(final Task task) {
        if (null == task.getValue()) {
            throw new ValidationException("Task value cannot be null");
        }
        this.validateTaskBasics(task);
        this.validateTaskDates(task);
        this.validateTaskAssignments(task);
    }

    private void validateTaskBasics(final Task task) {
        final TaskValue value = task.getValue();
        if (null == value.name() || value.name().trim().isEmpty()) {
            throw new ValidationException("Task name is required");
        }
        if (null == value.status()) {
            throw new ValidationException("Task status is required");
        }
    }

    private void validateTaskDates(final Task task) {
        final TaskValue value = task.getValue();
        if (null != value.dueDate() && null != value.plannedStartDate()
            && value.dueDate().isBefore(value.plannedStartDate())) {
            throw new ValidationException("Task due date cannot be before planned start date");
        }
    }

    private void validateTaskAssignments(final Task task) {
        final TaskValue value = task.getValue();
        if (value.isStarted() && null == value.assignee()) {
            throw new ValidationException("Started task must have an assignee");
        }
    }
}

File: src/main/kotlin/com/projecthub/domain/task/validation/TaskValidator.kt
package com.projecthub.domain.task.validation

import com.projecthub.domain.validation.ValidationResult
import com.projecthub.domain.validation.Validator
import com.projecthub.domain.task.Task
import org.springframework.stereotype.Component

@Component
class TaskValidator : Validator<Task> {

    fun validateTasks(tasks: Set<Task>): ValidationResult {
        return ValidationResult.Builder().apply {
            tasks.forEach { task ->
                validateTaskDependencies(task, tasks)
            }
        }.build()
    }

    private fun ValidationResult.Builder.validateTaskDependencies(task: Task, allTasks: Set<Task>) {
        // Add dependency validation logic here when implementing task dependencies
        // This will be useful when we migrate the dependency features
    }

    fun validateCreate(task: Task): ValidationResult {
        return ValidationResult.Builder().apply {
            validateBasics(task)
            validateDates(task)
            validateAssignments(task)
        }.build()
    }

    private fun ValidationResult.Builder.validateBasics(task: Task) {
        if (task.name.isBlank()) {
            addError("Task name is required")
        }
    }

    private fun ValidationResult.Builder.validateDates(task: Task) {
        task.plannedStartDate?.let { start ->
            task.dueDate?.let { due ->
                if (due.isBefore(start)) {
                    addError("Task due date cannot be before planned start date")
                }
            }
        }
    }

    private fun ValidationResult.Builder.validateAssignments(task: Task) {
        if (task.status.isStarted && task.assigneeId == null) {
            addError("Started task must have an assignee")
        }
    }
}